File: src/rl_stack/domain/rewards.py
"""Pure reward signal primitives. No I/O, deterministic, replayable.

Each reward function takes (TaskSpec, TrajectoryRecord) and returns a
RewardSignal: a named contribution with a value in [-1, 1] and a short
explanation. A CompositeRewardPipeline combines signals with weights and
produces the final RewardRecord + a list of per-signal breakdowns (stored
on the trajectory as summaries for now; a dedicated table lands when the
rubric registry grows up).
"""
from __future__ import annotations

import json
from collections.abc import Callable
from dataclasses import dataclass

from .models import RewardPenalty, RewardRecord, TaskSpec, TrajectoryRecord


@dataclass(frozen=True)
class RewardSignal:
    name: str
    value: float
    reason: str
    weight: float = 1.0


RewardFn = Callable[[TaskSpec, TrajectoryRecord], RewardSignal]


def step_count_signal(task: TaskSpec, trajectory: TrajectoryRecord, *, weight: float = 0.3) -> RewardSignal:
    step_count = len(trajectory.steps)
    value = min(1.0, step_count / max(task.horizon * 2, 1))
    return RewardSignal(
        name="step_count",
        value=value,
        reason=f"{step_count} trajectory steps against horizon={task.horizon}",
        weight=weight,
    )


def error_penalty_signal(task: TaskSpec, trajectory: TrajectoryRecord, *, weight: float = 1.0) -> RewardSignal:
    if not trajectory.errors:
        return RewardSignal(name="error_penalty", value=0.0, reason="no errors", weight=weight)
    return RewardSignal(
        name="error_penalty",
        value=-1.0,
        reason=f"{len(trajectory.errors)} errors: {trajectory.errors[0][:120]}",
        weight=weight,
    )


def success_criteria_signal(
    task: TaskSpec, trajectory: TrajectoryRecord, *, weight: float = 0.7
) -> RewardSignal:
    if not task.success_criteria:
        return RewardSignal(
            name="success_criteria", value=0.0, reason="no criteria specified", weight=weight,
        )
    haystack = _trajectory_text(trajectory).lower()
    matches = [c for c in task.success_criteria if c.strip().lower() in haystack]
    ratio = len(matches) / len(task.success_criteria)
    return RewardSignal(
        name="success_criteria",
        value=ratio,
        reason=f"matched {len(matches)}/{len(task.success_criteria)}: {matches}",
        weight=weight,
    )


def finish_signal(task: TaskSpec, trajectory: TrajectoryRecord, *, weight: float = 0.3) -> RewardSignal:
    for step in reversed(trajectory.steps):
        if step.actor != "policy":
            continue
        payload = _maybe_json(step.content)
        if payload and (payload.get("tool") or payload.get("name")) == "finish":
            return RewardSignal(
                name="finish", value=1.0, reason="policy called finish", weight=weight,
            )
        break
    return RewardSignal(name="finish", value=0.0, reason="no finish call", weight=weight)


def tests_pass_signal(
    task: TaskSpec, trajectory: TrajectoryRecord, *, weight: float = 0.8
) -> RewardSignal:
    """Scan tool observations for the last `pytest`/`python -m pytest` invocation
    and score by returncode. Neutral (0) if the policy never ran tests."""
    last_result: dict | None = None
    for step in trajectory.steps:
        if step.actor != "environment":
            continue
        payload = _maybe_json(step.content)
        if not payload or payload.get("tool") != "run_command":
            continue
        command = str(payload.get("command", ""))
        first = command.split(" ", 1)[0] if command else ""
        if first == "pytest" or command.startswith("python -m pytest") or command.startswith("python3 -m pytest"):
            last_result = payload
    if last_result is None:
        return RewardSignal(
            name="tests_pass", value=0.0, reason="no test run observed", weight=weight,
        )
    rc = last_result.get("returncode", 1)
    if rc == 0:
        return RewardSignal(
            name="tests_pass", value=1.0,
            reason=f"pytest exit 0 (sandbox={last_result.get('sandbox', 'unknown')})",
            weight=weight,
        )
    return RewardSignal(
        name="tests_pass", value=-1.0,
        reason=f"pytest exit {rc}",
        weight=weight,
    )


@dataclass(frozen=True)
class RubricSpec:
    name: str
    version: str
    signals: tuple[RewardFn, ...]

    @property
    def provenance(self) -> str:
        return f"{self.name}-{self.version}"


HEURISTIC_V1 = RubricSpec(
    name="heuristic",
    version="v1",
    signals=(
        step_count_signal,
        error_penalty_signal,
        success_criteria_signal,
        finish_signal,
    ),
)


CODING_V1 = RubricSpec(
    name="coding",
    version="v1",
    signals=(
        error_penalty_signal,
        success_criteria_signal,
        finish_signal,
        tests_pass_signal,
    ),
)


def score(task: TaskSpec, trajectory: TrajectoryRecord, rubric: RubricSpec) -> RewardRecord:
    """Run a rubric over a trajectory. Pure; safe to call in replay."""
    signals = [fn(task, trajectory) for fn in rubric.signals]
    weighted = [s.value * s.weight for s in signals]
    total_weight = sum(abs(s.weight) for s in signals) or 1.0
    terminal = max(-1.0, min(1.0, sum(weighted) / total_weight))

    penalties = [
        RewardPenalty(code=s.name, value=s.value * s.weight, reason=s.reason)
        for s in signals
        if s.value < 0
    ]
    audit_flags: list[str] = []
    if trajectory.errors:
        audit_flags.append("requires-manual-audit")
    if any(s.name == "success_criteria" and s.value == 0 for s in signals):
        audit_flags.append("no-success-criteria-match")

    step_count = max(len(trajectory.steps), 1)
    return RewardRecord(
        trajectory_id=trajectory.id,
        terminal_reward=round(terminal, 4),
        step_rewards=[round(terminal / step_count, 4)] * step_count,
        penalties=penalties,
        audit_flags=audit_flags,
        provenance=_encode_provenance(rubric, signals),
    )


def _encode_provenance(rubric: RubricSpec, signals: list[RewardSignal]) -> str:
    payload = {
        "rubric": rubric.provenance,
        "signals": [
            {"name": s.name, "value": round(s.value, 4), "weight": s.weight, "reason": s.reason}
            for s in signals
        ],
    }
    return json.dumps(payload)


def _trajectory_text(trajectory: TrajectoryRecord) -> str:
    return "\n".join(step.content for step in trajectory.steps)


def _maybe_json(text: str) -> dict | None:
    try:
        value = json.loads(text)
    except (json.JSONDecodeError, TypeError):
        return None
    return value if isinstance(value, dict) else None

File: crates/horizon-core/src/rewards.rs
//! Pure reward signal primitives. No I/O, deterministic, replayable.
//!
//! These heuristic rubrics exist for replay and for the in-house repo
//! path. Production reward for long-horizon agents comes from verifiers
//! rubrics or judge models; do not extend these with more string matching.

use serde_json::Value;

use crate::models::{RewardRecord, RewardSignal, TaskSpec, Trajectory};
use crate::round_to;

fn signal(name: &str, value: f64, reason: impl Into<String>, weight: f64) -> RewardSignal {
    RewardSignal {
        name: name.into(),
        value,
        weight,
        reason: reason.into(),
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SignalKind {
    StepCount,
    ErrorPenalty,
    SuccessCriteria,
    Finish,
    TestsPass,
}

impl SignalKind {
    pub fn name(self) -> &'static str {
        match self {
            SignalKind::StepCount => "step_count",
            SignalKind::ErrorPenalty => "error_penalty",
            SignalKind::SuccessCriteria => "success_criteria",
            SignalKind::Finish => "finish",
            SignalKind::TestsPass => "tests_pass",
        }
    }

    pub fn default_weight(self) -> f64 {
        match self {
            SignalKind::StepCount => 0.3,
            SignalKind::ErrorPenalty => 1.0,
            SignalKind::SuccessCriteria => 0.7,
            SignalKind::Finish => 0.3,
            SignalKind::TestsPass => 0.8,
        }
    }

    pub fn evaluate(self, task: &TaskSpec, trajectory: &Trajectory, weight: f64) -> RewardSignal {
        match self {
            SignalKind::StepCount => {
                let step_count = trajectory.steps.len();
                let denom = (task.horizon as usize * 2).max(1);
                let value = (step_count as f64 / denom as f64).min(1.0);
                signal(
                    "step_count",
                    value,
                    format!("{step_count} steps against horizon {}", task.horizon),
                    weight,
                )
            }
            SignalKind::ErrorPenalty => {
                if trajectory.errors.is_empty() {
                    return signal("error_penalty", 0.0, "no errors", weight);
                }
                let first: String = trajectory.errors[0].chars().take(120).collect();
                signal(
                    "error_penalty",
                    -1.0,
                    format!("{} errors: {first}", trajectory.errors.len()),
                    weight,
                )
            }
            SignalKind::SuccessCriteria => {
                if task.success_criteria.is_empty() {
                    return signal("success_criteria", 0.0, "no criteria specified", weight);
                }
                let haystack = trajectory_text(trajectory).to_lowercase();
                let matches: Vec<&str> = task
                    .success_criteria
                    .iter()
                    .filter(|c| haystack.contains(&c.trim().to_lowercase()))
                    .map(String::as_str)
                    .collect();
                let ratio = matches.len() as f64 / task.success_criteria.len() as f64;
                let reason = format!(
                    "matched {}/{}: {}",
                    matches.len(),
                    task.success_criteria.len(),
                    matches.join(", ")
                );
                signal("success_criteria", ratio, reason, weight)
            }
            SignalKind::Finish => {
                for step in trajectory.steps.iter().rev() {
                    if step.actor != "policy" {
                        continue;
                    }
                    if crate::tools::is_finish(&step.content) {
                        return signal("finish", 1.0, "policy called finish", weight);
                    }
                    break;
                }
                signal("finish", 0.0, "no finish call", weight)
            }
            SignalKind::TestsPass => tests_pass(trajectory, weight),
        }
    }
}

fn tests_pass(trajectory: &Trajectory, weight: f64) -> RewardSignal {
    let mut last: Option<Value> = None;
    for step in &trajectory.steps {
        if step.actor != "environment" {
            continue;
        }
        let Some(payload) = maybe_json(&step.content) else {
            continue;
        };
        if payload.get("tool").and_then(Value::as_str) != Some("run_command") {
            continue;
        }
        let command = payload.get("command").and_then(Value::as_str).unwrap_or("");
        let first = command.split(' ').next().unwrap_or("");
        if first == "pytest"
            || command.starts_with("python -m pytest")
            || command.starts_with("python3 -m pytest")
        {
            last = Some(payload);
        }
    }
    let Some(last) = last else {
        return signal("tests_pass", 0.0, "no test run observed", weight);
    };
    let rc = last.get("returncode").and_then(Value::as_i64).unwrap_or(1);
    if rc == 0 {
        let sandbox = last
            .get("sandbox")
            .and_then(Value::as_str)
            .unwrap_or("unknown");
        return signal(
            "tests_pass",
            1.0,
            format!("pytest exit 0 (sandbox={sandbox})"),
            weight,
        );
    }
    signal("tests_pass", -1.0, format!("pytest exit {rc}"), weight)
}

#[derive(Debug, Clone, PartialEq)]
pub struct RubricSpec {
    pub name: &'static str,
    pub version: &'static str,
    pub signals: Vec<(SignalKind, f64)>,
}

impl RubricSpec {
    pub fn provenance(&self) -> String {
        format!("{}-{}", self.name, self.version)
    }

    pub fn signal_names(&self) -> Vec<&'static str> {
        self.signals.iter().map(|(k, _)| k.name()).collect()
    }
}

fn with_default(kind: SignalKind) -> (SignalKind, f64) {
    (kind, kind.default_weight())
}

pub fn heuristic_v1() -> RubricSpec {
    RubricSpec {
        name: "heuristic",
        version: "v1",
        signals: vec![
            with_default(SignalKind::StepCount),
            with_default(SignalKind::ErrorPenalty),
            with_default(SignalKind::SuccessCriteria),
            with_default(SignalKind::Finish),
        ],
    }
}

pub fn coding_v1() -> RubricSpec {
    RubricSpec {
        name: "coding",
        version: "v1",
        signals: vec![
            with_default(SignalKind::ErrorPenalty),
            with_default(SignalKind::SuccessCriteria),
            with_default(SignalKind::Finish),
            with_default(SignalKind::TestsPass),
        ],
    }
}

pub fn strict_finish_v1() -> RubricSpec {
    RubricSpec {
        name: "strict-finish",
        version: "v1",
        signals: vec![
            (SignalKind::ErrorPenalty, 1.0),
            (SignalKind::Finish, 1.5),
            (SignalKind::SuccessCriteria, 0.5),
        ],
    }
}

/// Central rubric registry, keyed by provenance (`heuristic-v1`).
pub fn rubrics() -> Vec<RubricSpec> {
    vec![heuristic_v1(), coding_v1(), strict_finish_v1()]
}

pub fn get_rubric(name: &str) -> Option<RubricSpec> {
    rubrics().into_iter().find(|r| r.provenance() == name)
}

pub fn rubric_names() -> Vec<String> {
    let mut names: Vec<String> = rubrics().iter().map(RubricSpec::provenance).collect();
    names.sort();
    names
}

/// Run a rubric over a trajectory. Pure; safe to call in replay.
pub fn score(task: &TaskSpec, trajectory: &Trajectory, rubric: &RubricSpec) -> RewardRecord {
    let signals: Vec<RewardSignal> = rubric
        .signals
        .iter()
        .map(|(kind, weight)| kind.evaluate(task, trajectory, *weight))
        .collect();
    let weighted: f64 = signals.iter().map(|s| s.value * s.weight).sum();
    let total_weight: f64 = {
        let w: f64 = signals.iter().map(|s| s.weight.abs()).sum();
        if w == 0.0 {
            1.0
        } else {
            w
        }
    };
    let terminal = (weighted / total_weight).clamp(-1.0, 1.0);
    let mut audit_flags = Vec::new();
    if !trajectory.errors.is_empty() {
        audit_flags.push("requires-manual-audit".to_string());
    }
    if signals
        .iter()
        .any(|s| s.name == "success_criteria" && s.value == 0.0)
    {
        audit_flags.push("no-success-criteria-match".to_string());
    }
    RewardRecord {
        terminal_reward: round_to(terminal, 4),
        rubric: rubric.provenance(),
        source: "rubric".into(),
        signals: signals
            .into_iter()
            .map(|s| RewardSignal {
                value: round_to(s.value, 4),
                ..s
            })
            .collect(),
        audit_flags,
    }
}

fn trajectory_text(trajectory: &Trajectory) -> String {
    trajectory
        .steps
        .iter()
        .map(|s| s.content.as_str())
        .collect::<Vec<_>>()
        .join("\n")
}

fn maybe_json(text: &str) -> Option<Value> {
    let v: Value = serde_json::from_str(text).ok()?;
    if v.is_object() {
        Some(v)
    } else {
        None
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::TrajectoryStep;

    fn task() -> TaskSpec {
        TaskSpec {
            id: "task-1".into(),
            prompt: "do it".into(),
            repo_snapshot: ".".into(),
            horizon: 4,
            success_criteria: vec!["README".into()],
        }
    }

    #[test]
    fn finish_and_criteria_score_positive() {
        let steps = vec![
            TrajectoryStep::new(
                0,
                "policy",
                "action",
                r#"{"tool":"read_file","input":{"path":"README.md"}}"#,
            ),
            TrajectoryStep::new(
                1,
                "environment",
                "observation",
                r#"{"tool":"read_file","content":"readme text"}"#,
            ),
            TrajectoryStep::new(
                2,
                "policy",
                "action",
                r#"{"tool":"finish","input":{"summary":"done"}}"#,
            ),
        ];
        let traj = Trajectory {
            steps,
            errors: vec![],
        };
        let reward = score(&task(), &traj, &heuristic_v1());
        assert!(
            reward.terminal_reward > 0.4,
            "got {}",
            reward.terminal_reward
        );
        assert!(reward.signals.iter().all(|s| s.value >= 0.0));
        assert_eq!(reward.rubric, "heuristic-v1");
        assert_eq!(reward.source, "rubric");
    }

    #[test]
    fn errors_penalize_and_flag() {
        let traj = Trajectory {
            steps: vec![],
            errors: vec!["boom".into()],
        };
        let reward = score(&task(), &traj, &heuristic_v1());
        assert!(reward.terminal_reward < 0.0);
        assert!(reward
            .signals
            .iter()
            .any(|s| s.name == "error_penalty" && s.value < 0.0));
        assert!(reward
            .audit_flags
            .contains(&"requires-manual-audit".to_string()));
    }

    #[test]
    fn registry_lookup() {
        assert!(get_rubric("coding-v1").is_some());
        assert!(get_rubric("nope").is_none());
        assert_eq!(
            rubric_names(),
            vec!["coding-v1", "heuristic-v1", "strict-finish-v1"]
        );
    }
}
